vine/std/numeric/Nat.vi

```diff
@@ -0,0 +1,301 @@
+            };
+
+            if !~trim {
+              result.push_back(diff);
+            }
+            if diff != 0 {
+              ~trim = false;
+            }
+            continue;
+          }
+          (None, Some(b_num)) {
+            // underflow
+            return Nat::zero;
+          }
+          (Some(a_num), None) {
+            let diff = a_num - borrow as N32;
+            if !~trim {
+              result.push_back(diff);
+            }
+            if diff != 0 {
+              ~trim = false;
+            }
+
+            borrow = a_num < borrow as N32;
+            a
+          }
+          (None, None) { [] }
+        }
+      };
+      loop {
+        when {
+          !borrow {
+            result ++= remaining;
+            break;
+          }
+          remaining.pop_front() is Some(num) {
+            let diff = num - borrow as N32;
+            if !~trim {
+              result.push_back(diff);
+            }
+            if diff != 0 {
+              ~trim = false;
+            }
+            let borrow = num == 0;
+            continue;
+          }
+          _ {
+            // Underflow
+            return Nat::zero;
+          }
+        }
+      }
+      Nat(result)
+    }
+  }
+
+  pub impl mul_n32: Mul[Nat, N32, Nat] {
+    fn mul(Nat(a), n: N32) -> Nat {
+      if n == 0 {
+        return 0 as Nat;
+      }
+      let carry = 0;
+      let result = [];
+      for num in a.into_iter() {
+        let N64(lo, hi) = N64::mul_n32_n32(num, n);
+        result.push_back(lo + carry);
+        carry = hi;
+      }
+      if carry != 0 {
+        result.push_back(carry);
+      }
+      Nat(result)
+    }
+  }
+
+  pub impl mul: Mul[Nat, Nat, Nat] {
+    fn mul(a: Nat, b: Nat) -> Nat {
+      if a == Nat::zero || b == Nat::zero {
+        return Nat::zero;
+      }
+      let leading = [];
+      let current = Nat::zero;
+      for a_num in a!.into_iter() {
+        let prod = b * a_num;
+        current += prod;
+        leading.push_back(current!.pop_front().unwrap());
+      }
+      Nat(leading ++ current!)
+    }
+  }
+
+  pub impl div_n32: Div[Nat, N32, Nat] {
+    fn div(a: Nat, d: N32) -> Nat {
+      a.div_rem_n32(d).0
+    }
+  }
+
+  pub impl rem_n32: Rem[Nat, N32, N32] {
+    fn rem(a: Nat, d: N32) -> N32 {
+      a.div_rem_n32(d).1
+    }
+  }
+
+  pub fn .div_rem_n32(a: Nat, d: N32) -> (Nat, N32) {
+    if d == 0 {
+      debug::error("Division by zero");
+      unsafe::eraser
+    }
+
+    let nums = a!.reversed();
+
+    let Some(leading_num) = nums.pop_front() else {
+      return (Nat::zero, 0);
+    };
+    let q = [];
+    let (N64(qi, _), rem) = N64::div_rem_n32(leading_num as N64, d);
+    if qi != 0 {
+      q.push_front(qi);
+    }
+
+    for a_num in nums.into_iter() {
+      let cur = N64(a_num, rem);
+
+      let (qi, ri) = N64::div_rem_n32(cur, d);
+
+      q.push_front(qi as N32);
+      rem = ri;
+    }
+
+    (Nat(q), rem)
+  }
+
+  pub impl from_n32: Cast[N32, Nat] {
```

Comment:
Would also be nice to have `Cast[N64, Nat]`, `Cast[Nat, N32]`, and `Cast[Nat, N64]`.